### ISSUE TITLE
Setting QoS on the IotMqttPublish function block (IotMqtt_QoS1 vs IotMqtt_QoS0)

### ISSUE THREAD

**Issue by ghost** [br-automation-community/paho.mqtt.c-ar#34]
```
PublishClient(
      Enable := TRUE,
      IotMqttLink := MQTTClient.IotMqttLink,
      Topic := ADR(topic),
      QoS := IotMqtt_QoS1,
      Buffer := Buffer,
      BufferLength := Buffer_len,
      Retained := FALSE
);
```

Putting this QoS parameter to IotMqtt_QoS1, what does this actually do in the backend? if this is turned on, does the server send back a response which gets handled in the backend, to which the "Done" boolean is switched? I don't notice any difference when putting it to IotMqtt_QoS0.

Is the "Done" boolean switched based off of this QoS parameter, or should I look at "Info.DeliveriesCompleted"?

I am currently also subscribing on a self-fabricated response from the server, but if this works, I can scrap that part of my code.

**Comment by ghost** [comment]
Hello, 
QoS describes the behaviour in communication and is handled in the background of the stack. 
QoS = 0 means that the message is sent without any acknowledgement from the server, commonly known as "Fire And Forget".
QoS = 1 means that the sender keeps a copy of the message until the receiver is sending back an acknowledge. If this does not happen in a certain time, the sender resend the message. 
QoS = 2 means that the sender and the receiver use a kind of handshake to ensure that the package is only sent once, using flags to mark a resent package as duplicate. 
Please note, that the QoS can be renegotiated by a subscriber. The lowest QoS will then be used. 

You can find a nice explanation here: https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/

Basically all this is handled by the stack in the background. 
The "Done" flag is set to TRUE as soon as a response is registered on the send event from the stack. 
The "Info.DeliveriesCompleted" does show the number of acknowledged messages (delivery confirmed) when QoS > 0 but there is a message table in the background where you might get things mixed up, when publish many messages at once since all is handled asynchronous. 
What is the exact usecase so you would need this type of information? 

BR Fabian

**Comment by ghost** [comment]
I am wanting to send a message from a vehicle every 50 meters driven (I already know how to do this). So the messages will be sent at a maximum interval of about 6 - 7 seconds, as the vehicle had a max speed of 30kph. I want to verify that my message are arriving to my IAAS server (AWS). I have now written a program that does the checking of "Info.DeliveriesCompleted" really well. This was also way easier than using another subscriber.

I think the fact that the messages are handled asynchronously doesnt really matter in my usecase as I am planning to send messages fairly "slow".

sadly AWS does not support QoS2.

**Comment by ghost** [comment]
Hi, 
glad to hear you could manage the task and it is working. QoS seems the right approach for the usecase. 
Using the given interval and the assumed implementation this should work just fine for usecase. 

BR Fabian
